### ISSUE TITLE
[Question] Does there exist ways to invoke convert to TLV file from scripting?

### ISSUE THREAD

**Issue by KaseyFarron** [opentoonz/opentoonz#3212]
This is a question I thought best to ask here. I'm looking for ways to convert other raster files (TIF/PNG) to OpenToonz (TLV/TPL) type outside of OpenToonz by scripting.

Is there a way to do this? I checked the manual for ToonzScript but it doesn't seem to list very much.

Thank you!

**Comment by WolfInABowl** [comment]
I've looked into this previously without much luck.
It definitely seems to be possible to some degree as Clip Studio Paint converts its regular raster images to tlv when exporting an OpenToonz scene file. As it requires the user to point to the OT installation directory, there likely some sort of command-line stuff going on. 
Unfortunately, whether or not this functionality is exposed to ToonzScript is something that doesn't seem to be documented.

**Comment by RodneyBaker** [comment]
There use to be some example scripts on the old Digital Video Toonz website.
There were at least two conversion scripts and my memory says there was one for converting from raster to vector and one for converting from raster (or was it vector) to TLV.  
Unfortunately, that site does not appear to be currently accessible... perhaps via wayback machine?

I downloaded the scripts (as others did also) so hopefully we can locate them and share them.

I did NOT have any luck in successfully running those scripts but it was early days of Opentoonz and a lot of old functions and features weren't working back then.

Yes, this should be possible and even moreso now that Opentoonz has command line scripting capability.

**Comment by RodneyBaker** [comment]
Found the page I remembered via Wayback machine but as I suspected the specific script desired isn't there:

https://web.archive.org/web/20160613213240/http://www.toonz.com/htm/support/Script.htm

**Comment by KaseyFarron** [comment]
@RodneyBaker Doh!

@WolfInABowl Oh! My C++ is poor but I looked at scriptbindings and discovered there is a binding called `ToonzRasterConverter()` but I'm not sure which way it goes.

https://github.com/opentoonz/opentoonz/blob/83e2a21487e8da22051c105b191d3f7ec9a429d3/toonz/sources/toonzlib/scriptbinding_toonz_raster_converter.cpp#L10

I could successfully reference it in the command line but I get a TypeError.

```
t = new ToonzRasterConverter();
t.colorMapped = true // not sure if this is needed
a = t.(new Level("C:\\OpenToonz stuff\\projects\\test\\drawings\\test\\B..tif"))
a.save("C:\\OpenToonz stuff\\projects\\test\\drawings\\test\\B.tlv");
```

I also noticed this!

``` C++
if (context->argumentCount() != 1)
    return context->throwError(
        "Expected one argument (a raster Level or a raster Image)");
  QScriptValue arg = context->argument(0);
```

So it must be possible after all!

**Comment by KaseyFarron** [comment]
@RodneyBaker Doh!

@WolfInABowl Oh! My C++ is poor but I looked at scriptbindings and discovered there is a binding called `ToonzRasterConverter()` but I'm not sure which way it goes.

https://github.com/opentoonz/opentoonz/blob/83e2a21487e8da22051c105b191d3f7ec9a429d3/toonz/sources/toonzlib/scriptbinding_toonz_raster_converter.cpp#L10

I could successfully reference it in the command line but I get a TypeError on line 3 as `t.ToonzRasterConverter` [undefined] is not a function. 

``` JavaScript
t = new ToonzRasterConverter();
t.colorMapped = true // not sure if this is needed
a = t.ToonzRasterConverter(new Level("C:\\OpenToonz stuff\\projects\\test\\drawings\\test\\B..tif"))
a.save("C:\\OpenToonz stuff\\projects\\test\\drawings\\test\\B.tlv");
```

I also noticed this!

``` C++
if (context->argumentCount() != 1)
    return context->throwError(
        "Expected one argument (a raster Level or a raster Image)");
  QScriptValue arg = context->argument(0);
```

So maybe it's possible after all!

**Comment by KaseyFarron** [comment]
I managed to get a TIF to convert to a Toonz Raster level via command line. I used this.

``` JS
a = new Image("C:\\OpenToonz stuff\\projects\\test\\drawings\\test_0001.tif");
r = new ToonzRasterConverter();
b = r.convert(a)
b.save("C:\\OpenToonz stuff\\projects\\test\\drawings\\NAME.tlv");
```

So you can use `ToonzRasterConvert()` constructor and then `convert` method.

Unfortunately it only produces a level with two styles (black & white) and the black style spills out, so it looks broken! But what I want is "Painted TLV from non AA source" like from the browser window. I don't see anymore scriptbindings, but I'll keep looking.

**Comment by RodneyBaker** [comment]
No, that isn't quite there but... excellent progress.  :)

This in my estimation is a downside of the current cleanup process (which is what Opentoonz uses to 'convert' from standard raster to toonz raster.  It's default palette is limited to black.  I would think we need to replace the default palette (or place something in the correct path) so that a color palette would be used instead.

**Comment by KaseyFarron** [comment]
@RodneyBaker Nono, what I want is a full color conversion of non AA raster images into ToonzRaster levels. You can do this by choosing convert in the file browser on raster images and selecting "Painted TLV from non AA source" and the palette styles will match those already in the original image.

**Comment by RodneyBaker** [comment]
Yes, sorry.  I followed that.  I was just adding the underlying core process used to 'convert' standard raster to TLV.

For full color cleanup/conversion the palette just gets expanded to more colors.

Aside:  The lack of some of this functionality in the core cleanup/conversion feature set is why I find the recent addition and improvement to 'Separate Colors' interesting.  To my naive way of thinking, it provides a foundation on which better cleanup/conversion can be built.

**Comment by KaseyFarron** [comment]
I found this comment...

https://github.com/opentoonz/opentoonz/blob/83e2a21487e8da22051c105b191d3f7ec9a429d3/toonz/sources/include/convert2tlv.h#L79-L86

Hmm! Was I not supposed to have found the ToonzRasterConverter()? Is it WIP?

Oh well! Hopefully this functionality can be added by somebody, because having to convert things over to TLV one by one is a big pain.

**Comment by RodneyBaker** [comment]
Ah, I was concentrating on the issue at hand of scripting.

I've some good news for conversion of multiple files if that is your primary goal.
In the Browser we can select multiple files and when we invoke Convert it will process through them all.

Again, sadly, conversion to TLV in this way is problematic for me but hopefully it won't be for you.
I'm beginning to think the issue with that is on my end because no one else appears to be experiencing it.

**Comment by RodneyBaker** [comment]
This query could use some additional research and development to gain more understanding of current conversion capability to include from command line and scripting.

Specifically if we can identify current process and shortfalls that will help.

@KaseyFarron Did you find any additional information beyond what is captured currently in this report?

**Comment by RodneyBaker** [comment]
I've updated labeling on this report to reflect the need for additional R&D and documentation of current scripting capabilities especially as related to conversion of raster (and vector) to TLV level format.

**Comment by RodneyBaker** [comment]
I've been playing with toonzscript lately and happened upon this report which I had forgotten about.
So glad that it was recorded here on github and for your insights documented here @KaseyFarron.

It may not address the full color issue but the straight conversion from raster to TLV is a keeper.  :)

Here's a no frills modification without error checking** adapted from the script provided above:
```js
a = new Image("F:\\input.0001.tif");
o = "F:\\converted.tlv"
print("Input file:  " + a);
r = new ToonzRasterConverter();
b = r.convert(a);
b.save(o);
print("Output file:  " + o);```

When typing that last print line I really should add an if exist/else output to inform the user if the save was successful or not.
But then I need to do the same for the input file's existence also.

I do plan to document things I've been encountering in toonzscript (both potentially useful and trivial).
While toonz script is certainly limited I am constantly suprised by what can be accomplished with it.

Who knows, by the time we update toonz script I might even know better how to use it.  :)
